### PR TITLE
adjustments on sale_order_import

### PR DESCRIPTION
**Pull request by mra-odoo**
- add team_id in import layout
- identify partner with phone number (phone/mobile in res.partner).  In case there are multiple partners with the same phone number, pick the first one
- in case there is no partner matching with the phone number, create one (you can assign some random default values to required fields as necessary for now)
- it seems the program does not set the warehouse correctly in SO.  Please check and fix the logic.
- clean up the code following the OCA guidelines
- split import_sale_data method into smaller blocks (i.e. max 30 lines per method) for better maintenace/clearer structure.  This method is currently too long

### CODE REVIEW

File: sale_order_import/wizard/import_sale_view.xml [hunk 1]
```diff
@@ -57,7 +38,7 @@
             </field>
         </record>
 
-        <record id="action_import_sale_order" model="ir.actions.act_window">
+        <record id="import_sale_action" model="ir.actions.act_window">
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale_view.xml]: The view_id of this record should be "import_sale_view_form" instead of "import_sale_view_form"

**Comment by TimLai125** [comment]
Furthermore, when I tried to import the csv file, "There is no invoiceable line." warning popped up.
I tried to debug the module and noticed that the `qty_to_invoice` of the newly created `sale.order.line` is 0.
Please check if there is any adjustments needed to be made.

File: sale_order_import/wizard/import_sale.py [hunk 2]
```diff
@@ -2,202 +2,436 @@
...
+        """Get partner dict"""
         if partner_value not in partner_dict.keys():
-            partner = self.env['res.partner'].search(
-                [('name', '=', partner_value),
+#             partner = self.env['res.partner'].search(
+#                 [('name', '=', partner_value),
+#                  ('active', '=', True)])
+            res_partner = self.env['res.partner']
+            partner = res_partner.search(
+                ['|', ('phone', '=', partner_value),
```
[the remark this answers is not in the document]
Comment (mra-odoo) [reply sale_order_import/wizard/import_sale.py]: Hi Tim,

Answer is:
2. (phone == partner_value OR mobile == partner_value) AND active == true

Regards,
Mustufa R

File: sale_order_import/wizard/import_sale_view.xml (hunk 1 above)
Comment (mra-odoo) [reply sale_order_import/wizard/import_sale_view.xml]: Fixed. It should be import_sale_view_form so we have fixed it.

**Comment by mra-odoo** [comment]
_Furthermore, when I tried to import the csv file, "There is no invoiceable line." warning popped up.
I tried to debug the module and noticed that the qty_to_invoice of the newly created sale.order.line is 0._

For above point if product is having set up with **Invoice Policy = Delivery qty** then it will give this error.

So can you please check and give us test cases how we can handle that situation... Because our code will only with if product is having **Invoice Policy = Ordered qty** ...

File: sale_order_import/wizard/import_sale.py [hunk 3]
```diff
@@ -223,167 +488,108 @@ def import_sale_data(self):
...
-            line_name = fields.index('Line Description')
-            price_unit = fields.index('Line Unit Price')
-            product_qty = fields.index('Line Qty')
-            taxes_id = fields.index('Line Tax')
-            notes = fields.index('Notes')
-            pricelist_id = fields.index('Pricelist')
-            warehouse_id = fields.index('Warehouse')
+            sheet_fields = next(csv_iterator)
+
+            order_group = sheet_fields.index('Group')
```
Comment (yostashiro) [inline sale_order_import/wizard/import_sale.py]: We should avoid getting internal error when the column is missing in the import file.

Perhaps make the line like `order_group = sheet_fields.index('Group') if 'Group' in sheet_fields else False` or show an error message, and do the same for the other fields as well.

**Comment by mra-odoo** [comment]
Improved https://github.com/rfhk/sts-custom/pull/25#pullrequestreview-82353353 -> Now it will validate number of columns of CSV and raise warning if any column missing.

File: sale_order_import/wizard/import_sale.py [hunk 4]
```diff
@@ -490,19 +508,100 @@ def import_sale_data(self):
...
+            else:
+                missing_columns.append('Group')
+
+            partner_name = False
+            if 'Customer' in sheet_fields:
+                partner_name = sheet_fields.index('Customer')
+            else:
+                missing_columns.append('Customer')
+
+            partner_id = False
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: Should it be `partner_tel` be more reasonable?

File: sale_order_import/wizard/import_sale.py [hunk 5]
```diff
@@ -490,19 +508,100 @@ def import_sale_data(self):
...
-            notes = sheet_fields.index('Notes')
-            pricelist_id = sheet_fields.index('Pricelist')
-            warehouse_id = sheet_fields.index('Warehouse')
-            team_id = sheet_fields.index('Team')
-            carrier_id = sheet_fields.index('Carrier')
+            #  column validation
+            missing_columns = list(set(FIELDS_TO_IMPORT) - set(sheet_fields))
+            if missing_columns:
+                raise Warning(
+                    _('Missing columns in sheet are: \n %s'
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: 'Following columns are missing: \n %s'

File: sale_order_import/wizard/import_sale.py [hunk 6]
```diff
@@ -490,19 +508,100 @@ def import_sale_data(self):
...
+
+            carrier_id = False
+            if 'Carrier' in sheet_fields:
+                carrier_id = sheet_fields.index('Carrier')
+            else:
+                missing_columns.append('Carrier')
+
+            if missing_columns:
+                raise Warning(
+                    _('Missing columns in sheet are: \n %s'
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: 'Following columns are missing: \n %s'

File: sale_order_import/wizard/import_sale.py [hunk 7]
```diff
@@ -124,8 +124,17 @@ def _get_order_value_dict(self, row, error_line_vals, partner_id,
         else:
             self._get_team_dict(team_value, team_dict, error_line_vals)
 
+        carrier_value = row[carrier_id].strip()
+        if not carrier_value:
+            error_line_vals['error_name'] = error_line_vals['error_name']\
+                + _('Carrier is empty!') + '\n'
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: For the all error messages, please apply the following format:
'Column "XXX" cannot be empty.' / 'Column "XXX" cannot be less than 0.'

File: sale_order_import/wizard/import_sale.py [hunk 8]
```diff
@@ -2,202 +2,480 @@
...
+                             product_dict, product_id_value, order_item_dict,
+                             qty, price_unit_value):
+        """Get order item dict"""
+        if not error_log_id:
+            name = row[line_name].strip()
+            product_data = self.env['product.product'].browse(
+                product_dict[product_id_value]
+            )  # odoo11
+            if not name:
+                name = product_data['value']['name']
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: If the "Line Description" of the CSV file is empty, Odoo Server Error will result,
```
Traceback (most recent call last):
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 646, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 307, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/home/tslai/ws/sst/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 339, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/tslai/ws/sst/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 332, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 927, in __call__
    return self.method(*args, **kw)
  File "/home/tslai/ws/sst/odoo/odoo/http.py", line 512, in response_wrap
    response = f(*args, **kw)
  File "/home/tslai/ws/sst/odoo/addons/web/controllers/main.py", line 928, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/tslai/ws/sst/odoo/addons/web/controllers/main.py", line 916, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/tslai/ws/sst/odoo/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/home/tslai/ws/sst/odoo/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/tslai/ws/sst/sts-custom/sale_order_import/wizard/import_sale.py", line 645, in import_sale_data
    qty, price_unit_value)
  File "/home/tslai/ws/sst/sts-custom/sale_order_import/wizard/import_sale.py", line 203, in _get_order_item_dict
    name = product_data['value']['name']
  File "/home/tslai/ws/sst/odoo/odoo/models.py", line 4657, in __getitem__
    return self._fields[key].__get__(self, type(self))
KeyError: 'value'
```

File: sale_order_import/wizard/import_sale.py [hunk 9]
```diff
@@ -2,202 +2,480 @@
...
+                              product_id, pricelist_id, warehouse_id,
+                              team_id, carrier_id, partner_dict, product_dict,
+                              pricelist_dict, picking_dict, warehouse_dict,
+                              team_dict, carrier_dict):
+        """Get order value dict"""
+        partner_value = row[partner_id].strip()
+
+        if not partner_value:
+            error_line_vals['error_name'] = error_line_vals['error_name']\
+                + _('Partner is empty!') + '\n'
```
Comment (TimLai125) [inline sale_order_import/wizard/import_sale.py]: For the all error messages, please apply the following format:
'Column "XXX" cannot be empty.' / 'Column "XXX" cannot be less than 0.'

**Comment by mra-odoo** [comment]
Hi Tim,

Changes suggested are improved, can you please review ?

Thank you,
Mustufa R

**Comment by TimLai125** [comment]
#28

**Pull request closed without merging**